Implement cancel button for the sidebar

tsmacdonald opened the pull request:
This cancel button (and its friends for emails and new pulses) closes the sidebar now:

![image](https://user-images.githubusercontent.com/784417/98604633-a8199b80-22a9-11eb-8ab7-8ce598255002.png)

tsmacdonald committed:
Implement cancel button for the sidebar

paulrosenzweig commented on frontend/src/metabase/dashboard/components/Dashboard.jsx:
```diff
@@ -228,6 +230,10 @@ export default class Dashboard extends Component {
   };
 
+  onCancel = () => {
+    this.setState({ isSharing: false });
```
It's not part of this PR, but `isSharing` might make more sense in Redux rather than component state. I think it hinges on whether we want to tie the sidebar behavior in with any other UI changes that are managed in redux. e.g. if we the sharing sidebar to auto-close when switching to editing mode, we can more cleanly handle that with reducers.

tsmacdonald replied:
@paulrosenzweig This sounds really appealing to me—there's already some funky behavior surrounding getting things closed. I spent a while exploring the actions, `Dashboard` state, etc. and didn't make too much headway. Do you have 10-15 minutes in the next few days to talk through this a little more?

robdaemon replied:
Ahhh cool I'd like a better explanation of this too - can I join the talk about it? I was looking at how other things were done like this using `props` instead of `state` but I couldn't figure out how that worked since `props` are supposed to be immutable?

There's clearly a redux-ism here I'm missing!

paulrosenzweig commented on frontend/src/metabase/sharing/components/SharingSidebar.jsx:
```diff
@@ -561,7 +561,7 @@ class SharingSidebar extends React.Component {
 
       return (
-        <Sidebar onCancel={() => true}>
+        <Sidebar onCancel={onCancel}>
```
👍 threading it through like this makes sense.

Separately, we might be able to consolidate the separate but identical `<Sidebar>` uses here, but that's outside of this PR.

paulrosenzweig reviewed (commented):
👏 woohoo first frontend PR on Metabase!

tsmacdonald committed:
Don't return duplicate dashboard subscriptions

robdaemon commented on src/metabase/models/pulse.clj:
```diff
@@ -240,24 +240,18 @@
+         query      (merge-with conj base-query (when dashboard_id
+                                                  {:join      [[PulseCard :pc]     [:= :p.id :pc.pulse_id]]
+                                                   :left-join [[DashboardCard :dc] [:= :pc.dashboard_card_id :dc.id]]
+                                                   :where     [:= :dc.dashboard_id dashboard_id]}))]
```
Does this actually merge properly, or do you end up with:

`[:and [:= :p.alert_condition nil]
         [:= :p.archived archived?]
 := :dc.dashboard_id dashboard_id]`

?

tsmacdonald replied:
@robdaemon Looks like you answered your own question re. `merge-with conj`. I confess that I don't understand SQL well enough to know why the `distinct` is necessary; I would've thought the query as-is wouldn't return duplicate results.

tsmacdonald committed:
Move sidebar cancelation into Redux

Add close icon

Fix popover behavior

tsmacdonald commented on frontend/src/metabase/dashboard/components/Dashboard.jsx:
```diff
@@ -228,8 +229,12 @@ export default class Dashboard extends Component {
+
   onSharingClick = () => {
-    this.setState({ isSharing: !this.state.isSharing });
+    this.props.setSharing(true);
```
I couldn't see any cases where the toggling was actually necessary

tsmacdonald commented on frontend/src/metabase/dashboard/components/Dashboard.jsx:
```diff
@@ -331,7 +337,11 @@ export default class Dashboard extends Component {
+                {...this.props}
+                isSharing={isSharing}
+                onCancel={this.onCancel}
+              />
```
Tbh it's still a little perverse to me that I need to thread both the state (`isSharing`) and its mutator (`onCancel`) through, but I can roll with it. I think it'd be more compelling if the data was more complex than a single boolean value.

tsmacdonald replied:
oh, d'oh, good call

tsmacdonald commented on frontend/src/metabase/dashboard/components/DashboardActions.jsx:
```diff
@@ -8,7 +8,7 @@ import PopoverWithTrigger from "metabase/components/PopoverWithTrigger";
 import Tooltip from "metabase/components/Tooltip";
 
-export const getDashboardActions = ({
+export const getDashboardActions = (self, {
```
I feel gross about this. I tried to use `getDashboardActions.bind(this)(this.props)` where this is used but it didn't work and I maybe don't grok `bind`

paulrosenzweig replied:
I agree on the gross feels. Could you extract the `PopoverWithTrigger` code below to its own class-based component?

paulrosenzweig reviewed (commented):
Looking good!

tsmacdonald commented:
Hey @paulrosenzweig, I addressed all your comments except refactoring `getDashboardActions` / the refs. I added that to a tech debt section on [the hub issue](https://github.com/metabase/metabase/issues/13671) and think it'd be more expedient to circle back to it once I've done a bit more of the feature work, because I think I'll do a better job with a little more experience, I'm worried about the feature work, and it's entirely possible the surrounding code will need to get changed as part of the feature work anyway.

So is it cool to merge this (into `subscribe-to-dashboards-oss`, not `master`)?

The build fails upstream too. Dealing with that is next on the list

paulrosenzweig reviewed (approved):
👍 merging now sounds good to me!

model2020 reviewed (commented):
Yes
